### DISCUSSION TITLE
Dark mode or dark theme and/or system theme support

### DISCUSSION THREAD

**Discussion by magicBeast** [alejandroautalan/pygubu-designer#195] [Ideas]
As all of my other apps run in dark mode, NOT having a dark mode/theme in an app is pretty bad, as it means getting one's retina's burnt out when using pygubu-designer :-) Please add a dark mode 🙏

**Comment by alejandroautalan** [comment]
Hello @magicBeast, thanks for trying pygubu.
I'm interested on fixing this in the future.

In the mean time, I can make some changes to make pygubu-designer look nice with some tkinter dark themes.

- What OS are you using?

Regards
Alejandro A.

**Reply by magicBeast** [reply]
I use both windows 11 and arch linux (btw ;-) ) - but using windows 11 at the moment. Thanks for the speedy reply. Quick question for you, apologies if I'm doing something silly; I thought I'd ask here instead as I'm new to using tkinter / customtkinter, instead of raising it as an issue yet. I'm trying to build a simple test app that uses _just_ customtkinter. I've used CTkToplevel at the root of the tree, and then CTkFrame, with some widgets inside that. It works fine in the preview inside pygubu-designer. However, if I add a command to a CTkButton widget, for example, calculate_button_clicked, it doesn't generate a method when I click generate under the application template.
    Also, when I run my app outside of pygubu-designer, it generates a small extra tk window, as well as the correct customtkinter only app window (picture attached below). I've tried with and without selecting 'Use ttk style definitions', though I don't believe this should be necessary for customtkinter.

generated app code:

#!/usr/bin/python3
import pathlib
import tkinter.ttk as ttk
import pygubu
PROJECT_PATH = pathlib.Path(__file__).parent
PROJECT_UI = PROJECT_PATH / "m-to-km.ui"


class MToKmApp:
    def __init__(self, master=None):
        self.builder = builder = pygubu.Builder()
        builder.add_resource_path(PROJECT_PATH)
        builder.add_from_file(PROJECT_UI)
        # Main widget
        self.mainwindow = builder.get_object("ctk_top_level", master)
        builder.connect_callbacks(self)

    def run(self):
        self.mainwindow.mainloop()


if __name__ == "__main__":
    app = MToKmApp()
    app.run()

generated .ui file, which only references customtkinter.

<?xml version='1.0' encoding='utf-8'?>
<interface version="1.3">
  <object class="customtkinter.CTkToplevel" id="ctk_top_level" named="True">
    <property name="minsize">300|125</property>
    <property name="title" translatable="yes">Mile to Km Converter</property>
    <containerlayout manager="grid">
      <property name="anchor">center</property>
    </containerlayout>
    <child>
      <object class="customtkinter.CTkFrame" id="ctkframe1">
        <layout manager="grid">
          <property name="column">0</property>
          <property name="padx">30</property>
          <property name="pady">30</property>
          <property name="row">0</property>
        </layout>
        <child>
          <object class="customtkinter.CTkEntry" id="user_input" named="True">
            <layout manager="grid">
              <property name="column">1</property>
              <property name="row">0</property>
            </layout>
          </object>
        </child>
        <child>
          <object class="customtkinter.CTkLabel" id="miles_label" named="True">
            <property name="text" translatable="yes">Miles</property>
            <layout manager="grid">
              <property name="column">2</property>
              <property name="row">0</property>
            </layout>
          </object>
        </child>
        <child>
          <object class="customtkinter.CTkLabel" id="is_equal_to_label" named="True">
            <property name="text" translatable="yes">is equal to</property>
            <layout manager="grid">
              <property name="column">0</property>
              <property name="row">1</property>
            </layout>
          </object>
        </child>
        <child>
          <object class="customtkinter.CTkLabel" id="result_label" named="True">
            <property name="text" translatable="yes">0</property>
            <layout manager="grid">
              <property name="column">1</property>
              <property name="row">1</property>
            </layout>
          </object>
        </child>
        <child>
          <object class="customtkinter.CTkLabel" id="km_label" named="True">
            <property name="padx">10</property>
            <property name="text" translatable="yes">Km</property>
            <layout manager="grid">
              <property name="column">2</property>
              <property name="row">1</property>
            </layout>
          </object>
        </child>
        <child>
          <object class="customtkinter.CTkButton" id="calculate_button" named="True">
            <property name="command" type="command" cbtype="simple">calculate_button_clicked</property>
            <property name="text" translatable="yes">Calculate</property>
            <layout manager="grid">
              <property name="column">1</property>
              <property name="pady">0 30</property>
              <property name="row">2</property>
            </layout>
          </object>
        </child>
      </object>
    </child>
  </object>
</interface>
![extra-tk-window](https://user-images.githubusercontent.com/16053168/222230434-8584e443-cae4-4d84-8caa-02a1957c360e.png)

**Comment by magicBeast** [comment]
Apologies, it didn't like me pasting the xml ui file in here, for obvious reasons :-)

**Comment by magicBeast** [comment]
![m-to-km ui](https://user-images.githubusercontent.com/16053168/222235384-b09771d8-51f8-4ece-a792-5fdd1f9ce84d.png)

**Comment by magicBeast** [comment]
![extra-tk-window](https://user-images.githubusercontent.com/16053168/222235717-3014505a-323d-4ff8-b257-745b0c4d9a18.png)

**Comment by alejandroautalan** [comment]
Hello @magicBeast.

About your customtkinter issues.  I have created a new release of pygubu and pygubu-designer fixing some glitches with customtkinter, please install the latest version.

To remove the extra window that you are viewing, use a customtkinter.CTk  widget as your main window.  I have re-created the basic examples [here](https://github.com/alejandroautalan/pygubu-designer/tree/master/examples/customtkinter).

Remember to install also the latest customtkinter version, because is the only version supported by the plugin.

If you have more problems, feel free to create a new issue or discussion, so the topics don't get mixed up.

**Comment by alejandroautalan** [comment]
About Dark mode, the latest pygubu-designer version adds support for some dark themes if they are installed.

Option 1 - SunValley themes

    pip install sv_ttk

Start pygubu-designer and go to Edit > Preferences > General > Preferred ttk theme, select "sun-valley-dark".

![2023-03-02_02-03-sv-dark](https://user-images.githubusercontent.com/3482471/222336396-79ceefc3-b6e3-4fe0-a6e8-27f846ffcf2a.png)


Option2 - ttkbootstrap themes

    pip install ttkbootstrap

Start pygubu-designer and go to Edit > Preferences > General > Preferred ttk theme, select "cyborg" or "darkly"

![2023-03-02_02-04-ttkb-cyborg](https://user-images.githubusercontent.com/3482471/222336500-16db83d8-6c1a-45d9-8d71-ff64782e47fe.png)

Try them and let me know.
Regards!

**Reply by empowersarch** [reply]
I noticed a UI scaling issue when I used the ttk themes sun-valley-dark and sun-valley-light on large monitors. The monitors are two Dell SE3223Q set to the recommended 3840 x 2160.

**UI Scaling works:**

- UI scaling works on pygubu-designer when using ttk themes, such as Cyborg, Darkly, and a few others I checked. 

**UI Scaling does not work:** 

- UI scaling doesn't work on pygubu-designer when I use the ttk themes sun-valley-dark or sun-valley-light.
- My monitors are set to scale 175%. I checked 150%, 125% , all three scales have the same issue.

**Comment by magicBeast** [comment]
I just can't get it to run at all now. I get this error:


![error](https://user-images.githubusercontent.com/16053168/222540167-aa3f1b1d-d4ff-4cc2-a0ae-f8eb7185a29b.png)

I removed the latest python (3.11) and did a fresh install, and installed the latest versions of customtkinter, pygubu, pygubu-designer, and sk-tkk using pip, as show here:



![pip-list](https://user-images.githubusercontent.com/16053168/222540453-e3ffabd3-511e-4f4d-a33e-baf26f18e52b.png)

I tried earlier to pull all of the packages from the latest commits directly from the github repos, just in-case any of the pypi versions weren't caught up yet, but that gave the same error :-(

**Reply by alejandroautalan** [reply]
It seems that the latest customtkinter is not enforcing the install of pillow on windows.
Install pillow:

    pip install Pillow

Let me know
Regards
Alejandro A.

**Comment by magicBeast** [comment]
Aha - that fixed it, thanks. Dark theme installed and working! I'll try out those examples for building a customtkinter only app tomorrow :-)

**Comment by magicBeast** [comment]
Hi Alejandro. I looked at those examples, but they are just on how to use customertkinter. I've already built several apps by hand and they work fine. My issue is when using pygubu-designer to graphically create an app. For example, if I want to create a simple demo app using normal ttk widgets, I can, under ttk, click Toplevel, click Frame, set it to grid, click button, preview it, and it works fine. Save out the .ui file, create some app code, then run it directly outside of pygubu-designer and it works fine, with a single window.
    I would expect to be able to do the same using the customerTkinter widgets. If I do the same, but using ONLY CTK widgets, like so:
Click CTkTopLevel, Click CTkFrame, set to grid layout, click CTkButton and then preview, it shows one window and *appears* to work exactly the same as before, but using CTk only widgets. However, when I run this code directly it creates an extra tk window (as shown in the picture above), even though I've never used one in the builder. So pygubu must be adding one as the root of the application. I've tried using CTk, instead of CTkToplevel as the root of the main window, but this results in even the preview not working. I'm saying that when CTkToplevel is selected pygubu should know not to create a normal tk window. Hope this is clear. If I'm building this incorrectly in the GUI let me know :-)
